visual_reasoning_nemo_collection/text/tokenizer.py

```diff
@@ -0,0 +1,56 @@
+# Copyright (C) IBM Corporation 2018
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+import torch
+from torch import nn
+from nemo.backends.pytorch import NonTrainableNM
+from nemo.core.neural_types import NeuralType, ChannelType, LengthsType
+from nemo.utils.decorators import add_port_docs
+from nemo.collections.nlp.data import WordTokenizer
+
+class Tokenizer(NonTrainableNM):
+    """
+    Wrapping `nemo.collections.nlp.data.WordTokenizer`:
+    - Tokenizing text directly to ids.
+    - Sorts the sentences in decreasing length order to make torch's `pack_padded_sequence` happy.
+    - Applies `torch.nn.utils.rnn.pad_sequence`.
+    - Outputs everything in tensor format.
+    """
+    @property
+    @add_port_docs()
+    def input_ports(self):
+        """Returns definitions of module input ports.
+        """
+        return {
+            "x": NeuralType(('B', 'T'), ChannelType()),
+        }
+
+    @property
+    @add_port_docs()
+    def output_ports(self):
+        """Returns definitions of module output ports.
+        """
+        return {
+            "output": NeuralType(('B', 'T'), ChannelType()),
+            "output_len": NeuralType(tuple('B'), LengthsType()),
+        }
+
+    def __init__(self, vocab_path: str):
+        super().__init__()
+
+        assert os.path.isfile(vocab_path)
+        self._tokenizer = WordTokenizer(vocab_path)
+
+    def forward(self, x):
+        tokenized = [self._tokenizer.text_to_ids(elem) for elem in x]
+        output_lens = [len(elem) for elem in tokenized]
+        # Sort in decreasing length order to make pack_padded_sequence happy
+        output_lens, tokenized = (list(e) for e in zip(*sorted(zip(output_lens, tokenized), reverse=True)))
+        tokenized = [torch.LongTensor(elem) for elem in tokenized]
+
+        tokenized_padded = nn.utils.rnn.pad_sequence(tokenized, batch_first=True, padding_value=self._tokenizer.vocab['<PAD>'])
+        tokenized_padded = torch.tensor(tokenized_padded, dtype=torch.long).to(self._device)
```

Comment:
Moreover, this components does two things: 
 - tokenization
 - indexing

In PTP I have "decoupled those two "functionalities" into two components...